### PR TITLE
add support for installing tinytex

### LINKED ISSUE #2788: Use tinytex to compile LaTeX documents
I think the **tinytex** package is mature enough now (it has been on CRAN for about half a year), so we may consider using it to compile LaTeX documents (including Rnw documents https://github.com/yihui/tinytex/issues/10). Note that the **rmarkdown** package has already used it to compile the intermediate `.tex` files to PDF in `rmarkdown::render()`. Some tips on how the IDE options should be passed to **tinytex**:

image

1. The main function to call is `tinytex::latexmk()`.

1. The LaTeX engine (`pdflatex`, `xelatex`, `lualatex`) is passed to the `engine` argument of `latexmk()`.

1. The `Clean auxiliary output` option is passed to the `clean` argument of `latexmk()`.

1. The `Enable shell escape` option is passed to `latexmk(engine_args = '-shell-escape')`.

The main advantages of using **tinytex** is that it will process bibliographies and resolve other cross-references correctly (compile the .tex for a correct number of times). If the user has installed [TinyTeX](https://yihui.name/tinytex/) (`tinytex::install_tinytex()`), missing LaTeX packages can also be installed automatically.

Let me know if you need any other information.

### PR DESCRIPTION
**Pull request by kevinushey**
Closes #2788.

This PR adds some UI for installing the `tinytex` package, and uses that to call `tinytex::install_tinytex()`.

Screen Shot 2019-09-30 at 4 38 44 PM

After running that, documents that need to be compiled to PDF (both `.Rnw` and `.Rmd`) should be able to be compiled successfully.

Some open thoughts:

1. Should we provide a modal dialog before starting the install? It could be a place where we could explain "what is tinytex?" and also provide a help link or similar.

2. We might also want to ensure this warning bar shows up when using an R Markdown document that will be compiled to PDF.

### CODE REVIEW

File: src/cpp/session/SessionModuleContext.cpp [hunk 1]
```diff
@@ -1070,6 +1070,25 @@ FilePath findProgram(const std::string& name)
...
 
+bool addTinytexToPathIfNecessary()
+{
+   if (isPdfLatexInstalled())
+      return false;
+
+   FilePath binPath;
+
+#if defined(_WIN32)
+   FilePath appData(core::system::getenv("APPDATA"));
```
Comment (jmcphers) [inline src/cpp/session/SessionModuleContext.cpp]: Is this guaranteed to be set correctly? (wondering if we should use `::SHGetKnownFolderPath`)

File: src/cpp/session/SessionModuleContext.cpp [hunk 2]
```diff
@@ -1070,6 +1070,25 @@ FilePath findProgram(const std::string& name)
...
+   binPath = appData.complete("TinyTeX\\bin\\win32");
+#elif defined(__APPLE__)
+   binPath = resolveAliasedPath("~/Library/TinyTeX/bin/x86_64-darwin");
```
Comment (jmcphers) [inline src/cpp/session/SessionModuleContext.cpp]: Any chance we could ask if the tinytex package itself where it installed pdflatex? (worried that these paths could drift in the future with updates to tinytex)

File: src/cpp/session/SessionModuleContext.cpp (hunk 1 above)
Comment (kevinushey) [reply src/cpp/session/SessionModuleContext.cpp]: Good idea -- and now that I think of it, you've already implemented this as part of the XDG support for config files! I'll make a change here.

File: src/cpp/session/SessionModuleContext.cpp (hunk 2 above)
Comment (kevinushey) [reply src/cpp/session/SessionModuleContext.cpp]: That's a good idea. I think we still want this fallback for the case where the tinytex package is not installed or not on the library paths for some reason, though. I'll make a change.

**Comment by jmcphers** [comment]
LGTM!

**Comment by coatless** [comment]
@kevinushey anyway that an alternative link to a true TeX distribution could be displayed as well? `tinytex` is a headache at times.

**Comment by kevinushey** [comment]
> `tinytex` is a headache at times.

I'd rather solve these issues in tinytex (assuming Yihui agrees it's something worth solving) than provide alternative download links, especially since those change and break over time.

What issues are you having with tinytex?

**Comment by coatless** [comment]
From [`pinp`](https://github.com/eddelbuettel/pinp)'s recent episode, `tinytex` ended up complicating the debugging process (@eddelbuettel can attest to the loss of a weekend). If someone is using a `.Rnw`, chances are they want more than what `tinytex` offers. For `rmarkdown`, that might be reversed... But, in this case, I'm more worried about `tinytex` infeccted student systems over a standardized LaTeX distribution. (That said, a majority of LaTeX work is now farmed out to [overleaf](https://overleaf.com).)

**Comment by kevinushey** [comment]
Was the issue that `tinytex` did not install some subset of latex packages required to render the document, or something else? AFAIK `tinytex` ultimately is just a portable TexLive distribution, but the selection of default packages may be too minimal.

**Comment by eddelbuettel** [comment]
It complete garbled an error message from latex, possibly interpreting it its own way and sending people scrambling the wrong way.  Including somewhat experienced folks like myself and Kurt.

I would prefer if I could opt out of tinytex, but I can't.  Not from R (b/c when rmarkdown et al depend on it there is no escaping), not from the package itself.  My ideal would be an option to toggle 'ok then, let me just pass this through' which would call the base R latex processor _just like it used to do_.   But I haven't had time to work on that / test it and won't for some time.  

In short, to me `tinytex` is pure nuisance.  I had all sorts of "fun" and "issues" and "bug" in 20+ years of R, but I never had latex issues (that weren't solvable at the distro level).
